src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs

```diff
@@ -29,44 +33,97 @@ private void EmitStackAllocInitializers(TypeSymbol type, BoundArrayInitializatio
                     _builder.EmitOpCode(ILOpCode.Ldsflda);
                     _builder.EmitToken(field, inits.Syntax, _diagnostics.DiagnosticBag);
                     _builder.EmitIntConstant(data.Length);
+                    _builder.EmitUnaligned(alignment: 1);
                     _builder.EmitOpCode(ILOpCode.Cpblk, -3);
-
-                    if (initializationStyle == ArrayInitializerStyle.Mixed)
+                }
+                else
+                {
+                    var syntaxNode = inits.Syntax;
+                    if (Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle, _diagnostics, syntax: syntaxNode, isOptional: true) is MethodSymbol createSpanHelper &&
+                        Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__get_Item, _diagnostics, syntax: syntaxNode, isOptional: true) is MethodSymbol spanGetItemDefinition)
                     {
-                        EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
+                        // Use RuntimeHelpers.CreateSpan and cpblk.
+                        var readOnlySpan = spanGetItemDefinition.ContainingType.Construct(elementType);
+                        Debug.Assert(TypeSymbol.Equals(readOnlySpan.OriginalDefinition, _module.Compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.ConsiderEverything));
+                        var spanGetItem = spanGetItemDefinition.AsMember(readOnlySpan);
+
+                        _builder.EmitOpCode(ILOpCode.Dup);
+
+                        // ldtoken <PrivateImplementationDetails>...
+                        // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType>(fldHandle)
+                        var field = _builder.module.GetFieldForData(data, alignment: (ushort)sizeInBytes, syntaxNode, _diagnostics.DiagnosticBag);
+                        _builder.EmitOpCode(ILOpCode.Ldtoken);
+                        _builder.EmitToken(field, syntaxNode, _diagnostics.DiagnosticBag);
+                        _builder.EmitOpCode(ILOpCode.Call, 0);
+                        var createSpanHelperReference = createSpanHelper.Construct(elementType).GetCciAdapter();
+                        _builder.EmitToken(createSpanHelperReference, syntaxNode, _diagnostics.DiagnosticBag);
+
+                        var temp = AllocateTemp(readOnlySpan, syntaxNode);
+                        _builder.EmitLocalStore(temp);
+                        _builder.EmitLocalAddress(temp);
+
+                        // span.get_Item[0]
+                        _builder.EmitIntConstant(0);
+                        _builder.EmitOpCode(ILOpCode.Call, 0);
+                        EmitSymbolToken(spanGetItem, syntaxNode, optArgList: null);
+
+                        _builder.EmitIntConstant(data.Length);
+                        if (sizeInBytes != 8)
```

Comment:
>if (sizeInBytes != 8) [](http://example.com/codeflow?start=24&length=21)

Looking at the `unaligned.` section of ECMA-335

>shall use unaligned. if the alignment is not known at compile time to be 8-byte.

From that, I infer that, if we align the RVA field type at 8 bytes, we will be compliant without the `unaligned` prefix. Should we use that option instead, rather than having conditional IL?

Reply:
Are there any downsides to conditional IL? On the other hand, it seems that aligning properly to 1/2/4/8 bytes instead of always to 8 bytes saves space.

Reply:
> Are there any downsides to conditional IL?

IL is bigger. Compiler's code is more complex. Test matrix is bigger, more scenarios need special attention. 

>  it seems that aligning properly to 1/2/4/8 bytes instead of always to 8 bytes saves space.

I do not think it actually does. At least not for the code emitted by our compilers because, I think, compiler aligns all blobs at 8 bytes regardless of the requested alignment, this is mentioned in one of the comments in code. The alignment that we explicitly set on the types is important for IL rewriting scenarios.

Reply:
> The alignment that we explicitly set on the types is important for IL rewriting scenarios.

I think this means it improves size for AOT and similar precompiled code (right?) which seems important as well.

Reply:
> I think this means it improves size for AOT and similar precompiled code (right?) which seems important as well.

To be honest, I do not know answer to these questions, And, even if that was the case, I am not sure if that is something that we should care about.

Reply:
> > The alignment that we explicitly set on the types is important for IL rewriting scenarios.

> I think this means it improves size for AOT and similar precompiled code (right?) which seems important as well.

It is required for functional correctness, to ensure that the code continues working after IL rewriting. Check https://github.com/dotnet/runtime/issues/79477 for examples of breaks caused by alignment issues.

Reply:
@jjonescz 

> It is required for functional correctness, to ensure that the code continues working after IL rewriting. It is not for performance.

I am interpreting this as: "it is unlikely that we get any complains if we trade (simplicity + IL size) for blob size."

Reply:
To clarify, I would prefer (simplicity + reduced IL size) over smaller blob size after some IL rewrite operation.

Reply:
@AlekseyTs
Sounds good. Your original comment mentioned the conditional IL. I have removed that and aligned the data to 8 bytes. However, for the other `cpblk` which is used for `byte`s only, I have left the 1-byte alignment and hence also `unaligned.1` - or did you want that to be aligned to 8 bytes as well?

Reply:
I do not think it makes sense to deoptimize the 1-byte case like that.

To make the two cases as similar as possible, you can emit `unaligned.1 cpblk` in both cases and emit the blob with `elementType.EnumUnderlyingTypeOrSelf().SpecialType.SizeInBytes()` alignment in both cases.

Reply:
> did you want that to be aligned to 8 bytes as well?

Yes, this is what I would do. I would align both at 8 bytes explicitly (we are aligning blobs at 8 bytes regardless), and would completely get rid of the `.unaligned` instruction.

Reply:
> I would align both at 8 bytes explicitly (we are aligning blobs at 8 bytes regardless), and would completely get rid of the .unaligned instruction.

It is size de-optimization for trimmer and AOT that are scenarios where we care about the side the most.

Reply:
> It is size de-optimization for trimmer and AOT that are scenarios where we care about the side the most. 

Perhaps I misinterpreted you first message on this discussion thread.

If the blob size optimization is important for AOT, then, I think we should keep the `.unaligned` usage on both code paths. We can get there by reverting the last commit. Sorry for the confusion.

Reply:
> Perhaps I misinterpreted you first message on this discussion thread.

I was commenting on why we set the alignment and why it would not work to omit it on the types. Sorry for the confusion.